NettyAdaptiveCumulator incorrectly releases `ByteBuf in` twice

<!-- Please answer these questions before submitting a bug report. -->

### What version of gRPC-Java are you using?

Version 1.51.0 introduced this issue.

### What is your environment?
Any

### What did you expect to see?

### What did you see instead?

### Steps to reproduce the bug


```java

io.netty.handler.codec.DecoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:294)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.local.LocalChannel.readInbound(LocalChannel.java:299)
	at io.netty.channel.local.LocalChannel.doBeginRead(LocalChannel.java:322)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.beginRead(AbstractChannel.java:834)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.read(DefaultChannelPipeline.java:1362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:835)
	at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:814)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.read(Http2ConnectionHandler.java:539)
	at io.netty.channel.AbstractChannelHandlerContext.invokeRead(AbstractChannelHandlerContext.java:839)
	at io.netty.channel.AbstractChannelHandlerContext.read(AbstractChannelHandlerContext.java:814)
	at io.netty.channel.DefaultChannelPipeline.read(DefaultChannelPipeline.java:1004)
	at io.netty.channel.AbstractChannel.read(AbstractChannel.java:290)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.readIfIsAutoRead(DefaultChannelPipeline.java:1422)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelReadComplete(DefaultChannelPipeline.java:1417)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:482)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:463)
	at io.netty.channel.DefaultChannelPipeline.fireChannelReadComplete(DefaultChannelPipeline.java:925)
	at io.netty.channel.local.LocalChannel.readInbound(LocalChannel.java:302)
	at io.netty.channel.local.LocalChannel.finishPeerRead0(LocalChannel.java:445)
	at io.netty.channel.local.LocalChannel.access$400(LocalChannel.java:50)
	at io.netty.channel.local.LocalChannel$5.run(LocalChannel.java:403)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.channel.DefaultEventLoop.run(DefaultEventLoop.java:54)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:83)
	at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:147)
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:101)
	at io.grpc.netty.NettyAdaptiveCumulator.cumulate(NettyAdaptiveCumulator.java:95)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:288)
	... 33 common frames omitted

```


In the case below, NettyAdaptiveCumulator incorrectly releases `ByteBuf in` twice.

* io.grpc.netty.NettyAdaptiveCumulator.cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in) 

* cumulation: UnpooledHeapByteBuf(ridx: 9, widx: 14, cap: 14/14) 
* in:  UnpooledHeapByteBuf(ridx: 0, widx: 8, cap: 8/8)

---

* The finally block of `NettyAdaptiveCumulator#mergeWithCompositeTail` is released once
* Release again in the finally block of `NettyAdaptiveCumulator#cumulate`

--- a/netty/src/main/java/io/grpc/netty/NettyAdaptiveCumulator.java
+++ b/netty/src/main/java/io/grpc/netty/NettyAdaptiveCumulator.java
@@ -204,11 +204,8 @@ class NettyAdaptiveCumulator implements Cumulator {
       composite.addFlattenedComponents(true, newTail);
       // New tail's ownership transferred to the composite buf.
       newTail = null;
-      in.release();
-      in = null;
-      // Restore the reader. In case it fails we restore the reader after releasing/forgetting
-      // the input and the new tail so that finally block can handles them properly.
       composite.readerIndex(prevReader);
+      in.release(); // Successful so took over ownership of in
     } finally {
       // If new tail's ownership isn't transferred to the composite buf.
       // Release it to prevent a leak.

--- a/netty/src/main/java/io/grpc/netty/NettyAdaptiveCumulator.java
+++ b/netty/src/main/java/io/grpc/netty/NettyAdaptiveCumulator.java
@@ -210,10 +210,6 @@ class NettyAdaptiveCumulator implements Cumulator {
       // the input and the new tail so that finally block can handles them properly.
       composite.readerIndex(prevReader);
     } finally {
-      // Input buffer was merged with the tail.
-      if (in != null) {
-        in.release();
-      }
       // If new tail's ownership isn't transferred to the composite buf.
       // Release it to prevent a leak.
       if (newTail != null) {

--- a/netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java
+++ b/netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java
@@ -533,13 +533,12 @@ public class NettyAdaptiveCumulatorTest {
 
       try {
         NettyAdaptiveCumulator.mergeWithCompositeTail(alloc, compositeThrows, in);
+        in = null; // On success it would be released
         fail("Cumulator didn't throw");
       } catch (UnsupportedOperationException actualError) {
         assertSame(expectedError, actualError);
         // Because of error, ownership shouldn't have changed so should not have been released.
-        int inRefCnt = in.refCnt();
-        in.release();
-        assertEquals(1, inRefCnt);
+        assertEquals(1, in.refCnt());
         // Tail released
         assertEquals(0, tail.refCnt());
         // Composite cumulation is retained
@@ -547,6 +546,9 @@ public class NettyAdaptiveCumulatorTest {
         // Composite cumulation loses the tail
         assertEquals(0, compositeThrows.numComponents());
       } finally {
+        if (in != null) {
+          in.release();
+        }
         compositeThrows.release();
       }
     }
@@ -571,13 +573,12 @@ public class NettyAdaptiveCumulatorTest {
 
       try {
         NettyAdaptiveCumulator.mergeWithCompositeTail(mockAlloc, compositeRo, in);
+        in = null; // On success it would be released
         fail("Cumulator didn't throw");
       } catch (UnsupportedOperationException actualError) {
-        int inRefCnt = in.refCnt();
-        in.release();
         assertSame(expectedError, actualError);
         // Because of error, ownership shouldn't have changed so should not have been released.
-        assertEquals(1, inRefCnt);
+        assertEquals(1, in.refCnt());
         // New buffer released
         assertEquals(0, newTail.refCnt());
         // Composite cumulation is retained
@@ -585,6 +586,9 @@ public class NettyAdaptiveCumulatorTest {
         // Composite cumulation loses the tail
         assertEquals(0, compositeRo.numComponents());
       } finally {
+        if (in != null) {
+          in.release();
+        }
         compositeRo.release();
       }
     }

--- a/netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java
+++ b/netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java
@@ -537,7 +537,9 @@ public class NettyAdaptiveCumulatorTest {
       } catch (UnsupportedOperationException actualError) {
         assertSame(expectedError, actualError);
         // Because of error, ownership shouldn't have changed so should not have been released.
-        assertEquals(1, in.refCnt());
+        int inRefCnt = in.refCnt();
+        in.release();
+        assertEquals(1, inRefCnt);
         // Tail released
         assertEquals(0, tail.refCnt());
         // Composite cumulation is retained
@@ -546,7 +548,6 @@ public class NettyAdaptiveCumulatorTest {
         assertEquals(0, compositeThrows.numComponents());
       } finally {
         compositeThrows.release();
-        in.release();
       }
     }
 
@@ -572,9 +573,11 @@ public class NettyAdaptiveCumulatorTest {
         NettyAdaptiveCumulator.mergeWithCompositeTail(mockAlloc, compositeRo, in);
         fail("Cumulator didn't throw");
       } catch (UnsupportedOperationException actualError) {
+        int inRefCnt = in.refCnt();
+        in.release();
         assertSame(expectedError, actualError);
         // Because of error, ownership shouldn't have changed so should not have been released.
-        assertEquals(1, in.refCnt());
+        assertEquals(1, inRefCnt);
         // New buffer released
         assertEquals(0, newTail.refCnt());
         // Composite cumulation is retained
@@ -583,7 +586,6 @@ public class NettyAdaptiveCumulatorTest {
         assertEquals(0, compositeRo.numComponents());
       } finally {
         compositeRo.release();
-        in.release();
       }
     }
   }

--- a/netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java
+++ b/netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java
@@ -536,7 +536,7 @@ public class NettyAdaptiveCumulatorTest {
         fail("Cumulator didn't throw");
       } catch (UnsupportedOperationException actualError) {
         assertSame(expectedError, actualError);
-        // Input must be released unless its ownership has been to the composite cumulation.
+        // Because of error, ownership shouldn't have changed so should not have been released.
         assertEquals(1, in.refCnt());
         // Tail released
         assertEquals(0, tail.refCnt());
@@ -573,7 +573,7 @@ public class NettyAdaptiveCumulatorTest {
         fail("Cumulator didn't throw");
       } catch (UnsupportedOperationException actualError) {
         assertSame(expectedError, actualError);
-        // Input must be released unless its ownership has been to the composite cumulation.
+        // Because of error, ownership shouldn't have changed so should not have been released.
         assertEquals(1, in.refCnt());
         // New buffer released
         assertEquals(0, newTail.refCnt());

--- a/netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java
+++ b/netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java
@@ -537,7 +537,7 @@ public class NettyAdaptiveCumulatorTest {
       } catch (UnsupportedOperationException actualError) {
         assertSame(expectedError, actualError);
         // Input must be released unless its ownership has been to the composite cumulation.
-        assertEquals(0, in.refCnt());
+        assertEquals(1, in.refCnt());
         // Tail released
         assertEquals(0, tail.refCnt());
         // Composite cumulation is retained
@@ -546,6 +546,7 @@ public class NettyAdaptiveCumulatorTest {
         assertEquals(0, compositeThrows.numComponents());
       } finally {
         compositeThrows.release();
+        in.release();
       }
     }
 
@@ -573,7 +574,7 @@ public class NettyAdaptiveCumulatorTest {
       } catch (UnsupportedOperationException actualError) {
         assertSame(expectedError, actualError);
         // Input must be released unless its ownership has been to the composite cumulation.
-        assertEquals(0, in.refCnt());
+        assertEquals(1, in.refCnt());
         // New buffer released
         assertEquals(0, newTail.refCnt());
         // Composite cumulation is retained
@@ -582,6 +583,7 @@ public class NettyAdaptiveCumulatorTest {
         assertEquals(0, compositeRo.numComponents());
       } finally {
         compositeRo.release();
+        in.release();
       }
     }
   }

